Sources/AsyncAlgorithms/Debounce/DebounceStorage.swift

```diff
@@ -0,0 +1,311 @@
+                                    break loop
+                                }
+                            }
+                        }
+
+                        do {
+                            try await group.waitForAll()
+                        } catch {
+                            // The upstream sequence threw an error
+                            let action = self.lock.withLock {
+                                self.stateMachine.upstreamThrew(error)
+                            }
+
+                            switch action {
+                            case .resumeContinuationWithErrorAndCancelTaskAndUpstreamContinuation(
+                                let downstreamContinuation,
+                                let error,
+                                let task,
+                                let upstreamContinuation,
+                                let clockContinuation
+                            ):
+                                upstreamContinuation?.resume(throwing: CancellationError())
+                                clockContinuation?.resume(throwing: CancellationError())
+
+                                task.cancel()
+
+                                downstreamContinuation.resume(throwing: error)
+
+                            case .cancelTaskAndClockContinuation(
+                                let task,
+                                let clockContinuation
+                            ):
+                                clockContinuation?.resume(throwing: CancellationError())
+                                task.cancel()
+
+                            case .none:
+                                break
+                            }
+
+                            group.cancelAll()
+                        }
+                    }
+                }
+
+                self.stateMachine.taskStarted(task)
+            }
+        }
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.lock.withLock { self.stateMachine.iteratorDeinitialized() }
+
+        switch action {
+        case .cancelTaskAndUpstreamAndClockContinuations(
+            let task,
+            let upstreamContinuation,
+            let clockContinuation
+        ):
+            upstreamContinuation?.resume(throwing: CancellationError())
+            clockContinuation?.resume(throwing: CancellationError())
+
+            task.cancel()
+
+        case .none:
+            break
+        }
+    }
+
+    func next() async rethrows -> Element? {
+        // We need to handle cancellation here because we are creating a continuation
+        // and because we need to cancel the `Task` we created to consume the upstream
+        return try await withTaskCancellationHandler {
+            // We always suspend since we can never return an element right away
+
+            self.lock.lock()
+            return try await withUnsafeThrowingContinuation { continuation in
```

Comment:
im VERY surprised this actually works as rethrowing. I would have expected that you need to use a `Result` to `_rethrowGet()` the value.

Reply:
I think this is related to the issue I've spotter here -> https://github.com/apple/swift/issues/60821.
We can bypass the mechanism by wrapping the call inside any rethrowing function (like `withTaskCancellationHandler`)

Reply:
yea that really seems like a compiler soundness hole to me. If at all possible we should avoid relying on it and paint everything into one common handler; it might be worth a secondary investigation of which approach; replicating this bug into a method to with a closure or using the rethrow-get both simulate a un-yet-implemented `rethrow(unchecked)` compiler feature.

Reply:
Yeah I found that soundness hole as well and wanted to check that actually. I just changed this to use `Result<Element?, Error>` instead.